docs/src/proposals/bidirectional_quic_reply.md

```diff
@@ -0,0 +1,100 @@
+bidirectional channel once and keep it alive to receive notifications about the transactions that it 
+has sent to the cluster. If the connection will be kept open by the server for a predetermined N 
+seconds and then will be closed. We can emit these notification at the most important stages of banking
+stage and where transactions are probably lost.
+
+A TPU client which connects to FANOUT_SLOTS/NUMBER_OF_SLOTS_PER_LEADER number of the next leader 
+then will initialize a bidirectional channel once with each of the leaders and send the transactions. 
+Each leader will then reply how it responded to the transactions and we can easily determine
+the lifecycle of the transaction. We add a mechanism in the TPU client called a bidirectional
+reply handler which will take care of the notifications received by each leader, and forward the 
+notifications to a crossbeam channel to be processed by the user of the TPU client.
+
+To avoid loops and over-notifications, each bidirectional channel will be limited to a single validator,
+the validator will not create bidirectional channels with other validators to get notifications.
+The size of the reply message will be a fixed length of N bytes so that the client will exactly 
+know that it has received a notification from the server after receiving N bytes. So the length of the 
+notification is limited to 128 bytes maximum. Here is the proposed structure:
+
+```
+pub struct QuicReplyMessage {
+    sender_identity: Pubkey,
+    transaction_signature: Signature,
+    message: [u8; 128],
+    approximate_slot: Slot, // slot when the validator was leader
+}
+```
+
+We can additionally add limits so that this feature could not be used as an attack vector.
+1. Number of connections in bidirectional stream 5 unstaked and 5 staked.
+2. Timeout for connection on server side : 30 seconds (i.e server will close connection in 30
+seconds)
+3. Additionally server will close the connection when client closes the connection.
+4. Timeout for notification message on client side : 10s
+5. Max number of transaction per connection : 10000
+
+
+## In working
+
+We have already implemented this proposal for test purpose in v1.14.
+https://github.com/solana-labs/solana/pull/29954
+
+We have done some tests using mango bencher on a private cluster:
+
+Example output:
+
+Here we have multiple messages represented as ("message (slot)").
+
+For successful transaction without congestion:
+```
+executed ([1147]), 
+```
+
+For successful transaction during congestion:
+```
+Transaction would exceed max Block Cost Limit ([1070]), 
+Account in use (GUg52LVs8adjcdsGYYEwH5hGWfthgknQai8Pk4GQ2pWz) ([1070]), 
+Account in use (3sWUkgYwyDLVnzLUGGX8Y33ph5iPG1WgUJWTpczf21QM) ([1070, 1071]),
+executed ([1071]),
+```
+
+For silently failed transactions 
+(transaction was silently discarded as excess packet in sigverify stage) : 
+```
+discarded as excess packet ([1110])
+```
```

Comment:
Can you address the plan on upgrade? Does the validator serve both unistream bidirection stream?

Reply:
I can add few lines to the proposal, but yeah validator will work as before. It will handle a unidirectional connection too. In this case we wont have any notifications.